### PR TITLE
Convert golang e2e test to unit tests

### PR DESCRIPTION
**Pull request by dmvolod**

### CODE REVIEW

**Comment by galderz** [comment]
Looks good to me. Just double check what @Crumby said about keeping some tests as unit tests and others as integration tests.

**Comment by galderz** [comment]
@Crumby @rigazilla Thoughts?

File: build/run-tests.sh [hunk 1]
```diff
@@ -5,4 +5,4 @@ KUBECONFIG=${1-openshift.local.clusterup/openshift-apiserver/admin.kubeconfig}
 echo "Using KUBECONFIG '$KUBECONFIG'"
 
 go clean -testcache ./test/e2e
-go test -v ./test/e2e -timeout 45m
+go test -v ./test/e2e -run TestExternalServiceWithAuth -timeout 45m
```
Comment (Crumby) [inline build/run-tests.sh]: This should be omitted from the PR so the CI doesn't run only one test.

Comment (dmvolod) [reply build/run-tests.sh]: Yes, sure. But we need to decide which set of the e2e test will keep for CI. What do you think about these?

Comment (Crumby) [reply build/run-tests.sh]: With unit tests in place we can get rid off some e2e execution heavy duplicates and keep executing all in e2e. See my suggestion:

**Remove:** TestSimple, TestNodeStartup, TestClusterFormationWithTls, TestContainerCPUUpdate, TestContainerMemoryUpdate, TestContainerJavaOptsUpdate

**Keep:** TestClusterFormation, TestCacheService, TestPernamentCache, TestCheckDataSurviveToShutdown, TestExternalService, TestExternalServiceWithAuth

Basically we can remove all the `*Update` tests. As far as we are concerned they have equivalent unit test at this moment. `TestSimple` isn't a test. `TestNodeStartup` is subtest of `TestClusterFormation` and `TestClusterFormationWithTLS` doesn't provide any more assurance as clustering uses different protocol meaning that TLS enabled doesn't affect it.

@rigazilla What do you think about this?

File: test/unit/main_test.go [hunk 2]
```diff
@@ -0,0 +1,567 @@
...
+	}
+	return nil, nil
+}
+
+func TestBasicConfiguration(t *testing.T) {
+	infinispan := basicInfinispan
+	r := testDefaultConfiguration(t, infinispan, basicRestMock)
+	sset := tstutil.GetStatefulSet(r.GetClient(), types.NamespacedName{operatorNamespace, operatorName})
+
+	assert.Equal(t, corev1.URIScheme("HTTP"), sset.Spec.Template.Spec.Containers[0].LivenessProbe.HTTPGet.Scheme, "Liveness probe Scheme not defined as expected")
```
Comment (Crumby) [inline test/unit/main_test.go]: Should this checks be part of `testDefaultConfiguration`?

File: build/run-tests.sh (hunk 1 above)
Comment (dmvolod) [reply build/run-tests.sh]: Thanks for response, @Crumby 
In case of the issue #311 we should need to keep one of the *Update tests with replicas=2 or create a new one. In other cases, the list looks very good.

File: test/unit/main_test.go (hunk 2 above)
Comment (dmvolod) [reply test/unit/main_test.go]: I will look into it. Thanks.

**Comment by Crumby** [comment]
@dmvolod I haven't noticed noticed a deletion presence in the unit tests. We should also have such when user deletes infinispan resource then then operator deletes all the related resource it created.

**Comment by dmvolod** [comment]
Thanks, @Crumby , will do this test.

File: test/unit/main_test.go (hunk 2 above)
Comment (dmvolod) [reply test/unit/main_test.go]: @Crumby looks like it's a correct test, as testDefaultConfiguration uses for both URI schema's: HTTP and HTTPS

**Comment by dmvolod** [comment]
@Crumby I've done some research and we can't create test for deletion as some of the resources are creating and deleting by foreign controllers but not fake client and we can just test them if will create/delete/update by yourself.

**Comment by Crumby** [comment]
@dmvolod Aren't we supposed to send the deletion request for some resources upon infinispan CR deletion? Eg. service, statefulsets, custom secrets...

**Comment by dmvolod** [comment]
> @dmvolod Aren't we supposed to send the deletion request for some resources upon infinispan CR deletion? Eg. service, statefulsets, custom secrets...

@Crumby no, all auto-generated resources are referenced by Infinispan CRD
`controllerutil.SetControllerReference(m, dep, r.scheme)`
